zap/src/parser/convert.rs

```diff
@@ -0,0 +1,336 @@
+		syntax_config
+			.decls
+			.iter()
+			.filter(|decl| matches!(decl, SyntaxDecl::Ty(_)))
+			.map(|decl| match decl {
+				SyntaxDecl::Ty(tydecl) => tydecl.name.name,
+				_ => unreachable!(),
+			})
+			.collect(),
+	);
+
+	let config = syntax_config.into_config(&mut state);
+
+	(config, state.into_daigs())
+}
+
+struct ConvertState<'src> {
+	daigs: Vec<Diagnostic<()>>,
+	tydecls: HashSet<&'src str>,
+}
+
+impl<'src> ConvertState<'src> {
+	fn new(tydecls: HashSet<&'src str>) -> Self {
+		Self {
+			daigs: Vec::new(),
+			tydecls,
+		}
+	}
+
+	fn push_diag(&mut self, diag: Diagnostic<()>) {
+		self.daigs.push(diag);
+	}
+
+	fn into_daigs(self) -> Vec<Diagnostic<()>> {
+		self.daigs
+	}
+
+	fn tydecl_exists(&self, name: &'src str) -> bool {
+		self.tydecls.contains(name)
+	}
+}
+
+trait IntoConfig<T> {
+	fn into_config(self, state: &mut ConvertState) -> T;
+}
+
+impl<'src> IntoConfig<Config<'src>> for SyntaxConfig<'src> {
+	fn into_config(self, state: &mut ConvertState) -> Config<'src> {
+		let mut write_checks = false;
+		let mut typescript = false;
+
+		let mut server_output = None;
+		let mut client_output = None;
+
+		let mut casing = Casing::Pascal;
+
+		for opt in self.opts {
+			match opt.name.into_config(state) {
+				"write_checks" => match opt.value.kind {
+					SyntaxOptValueKind::Bool(value) => write_checks = value.into_config(state),
+
+					_ => state.push_diag(
+						Diagnostic::warning()
+							.with_code("W0:001")
+							.with_message("invalid value for option, expected boolean")
+							.with_labels(vec![Label::primary((), opt.value.span())
+								.with_message("invalid value here, expected boolean")]),
+					),
+				},
+
+				"typescript" => match opt.value.kind {
+					SyntaxOptValueKind::Bool(value) => typescript = value.into_config(state),
+
+					_ => state.push_diag(
+						Diagnostic::warning()
+							.with_code("W0:001")
+							.with_message("invalid value for option, expected boolean")
+							.with_labels(vec![Label::primary((), opt.value.span())
+								.with_message("invalid value here, expected boolean")]),
+					),
+				},
+
+				"server_output" => match opt.value.kind {
+					SyntaxOptValueKind::Str(value) => server_output = Some(value.into_config(state)),
+
+					_ => state.push_diag(
+						Diagnostic::warning()
+							.with_code("W0:001")
+							.with_message("invalid value for option, expected string")
+							.with_labels(vec![Label::primary((), opt.value.span())
+								.with_message("invalid value here, expected string")]),
+					),
+				},
+
+				"client_output" => match opt.value.kind {
+					SyntaxOptValueKind::Str(value) => client_output = Some(value.into_config(state)),
+
+					_ => state.push_diag(
+						Diagnostic::warning()
+							.with_code("W0:001")
+							.with_message("invalid value for option, expected string")
+							.with_labels(vec![Label::primary((), opt.value.span())
+								.with_message("invalid value here, expected string")]),
+					),
+				},
```

Comment:
We should extract this similar logic into a function.
Maybe it should take in `SyntaxOptValueKind` and then creates the diagnostic if it doesn't fit in correctly.

Reply:
unfortunately I can't think of a way to implement that because these modify the variables around L66.

Reply:
Make a base diagnostic with the following before the match:

```rs
Diagnostic::warning()
	.with_code("W0:001")
	.with_message("invalid value for option, expected boolean")
```
and then add the labels to it. Like:
`invalid_value.with_labels`...